libvast/builtins/operators/from_read.cpp

```diff
@@ -0,0 +1,257 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/concept/parseable/string.hpp>
+#include <vast/concept/parseable/vast/identifier.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/error.hpp>
+#include <vast/legacy_pipeline.hpp>
+#include <vast/logger.hpp>
+#include <vast/plugin.hpp>
+
+#include <arrow/type.h>
+#include <caf/error.hpp>
+
+namespace vast::plugins::from_read {
+
+namespace {
+
+class from_operator final : public crtp_operator<from_operator> {
```

Comment:
Since this already uses polymorphism and one can be constructed with either loader or parser i would see it as two classes instead. One responsible from handling the loader and the other for handling the parser.

Reply:
Ideally `from_operator` would not care about the parser, and `read_operator` would not care about the loader. However, the `make_default_parser` and `make_default_loader` API expects an `operator_control_plane&`, which we only get when instantiating the already-built pipeline. Perhaps we can change the API to improve the situation.

Reply:
API was changed to improve this situation.